symbolic-base/src/ZkFold/Symbolic/Data/Ord.hs

```diff
@@ -98,40 +99,71 @@ getBitsBE x =
 bitwiseGE xs ys = Bool $
   symbolic2F xs ys
     (\us vs -> Par1 $ Haskell.bool zero one (toList us Haskell.>= toList vs))
-    $ \is js -> Par1 <$> blueprintGE is js
+    $ \is js -> Par1 <$> blueprintGE @r is js
 
-blueprintGE :: (MonadCircuit i a m, Z.Zip f, Foldable f) => f i -> f i -> m i
+blueprintGE :: forall r i a m f . (Arithmetic a, MonadCircuit i a m, Z.Zip f, Foldable f, KnownNat r) => f i -> f i -> m i
 blueprintGE xs ys = do
-  (_, hasNegOne) <- circuitDelta xs ys
+  (_, hasNegOne) <- circuitDelta @r xs ys
   newAssigned $ \p -> one - p hasNegOne
 
-bitwiseGT :: forall c f . (Symbolic c, Z.Zip f, Foldable f) => c f -> c f -> Bool c
+bitwiseGT :: forall r c f . (Symbolic c, Z.Zip f, Foldable f, KnownNat r) => c f -> c f -> Bool c
 -- ^ Given two lists of bits of equal length, compares them lexicographically.
 bitwiseGT xs ys = Bool $
   symbolic2F xs ys
     (\us vs -> Par1 $ Haskell.bool zero one (toList us Haskell.> toList vs))
     $ \is js -> do
-      (hasOne, hasNegOne) <- circuitDelta is js
+      (hasOne, hasNegOne) <- circuitDelta @r is js
       Par1 <$> newAssigned (\p -> p hasOne * (one - p hasNegOne))
 
-circuitDelta :: forall i a m f . (MonadCircuit i a m, Z.Zip f, Foldable f) => f i -> f i -> m (i, i)
+-- | Compare two sets of r-bit words lexicographically
+--
+circuitDelta :: forall r i a m f . (Arithmetic a, MonadCircuit i a m, Z.Zip f, Foldable f, KnownNat r) => f i -> f i -> m (i, i)
 circuitDelta l r = do
     z1 <- newAssigned (Haskell.const zero)
     z2 <- newAssigned (Haskell.const zero)
     foldM update (z1, z2) $ Z.zip l r
         where
+            bound = scale ((2 ^ value @r) -! 1) one
+
             -- | If @z1@ is set, there was an index i where @xs[i] == 1@ and @ys[i] == 0@ and @xs[j] == ys[j]@ for all j < i.
             -- In this case, no matter what bit states are after this index, @z1@ and @z2@ are not updated.
             --
             --   If @z2@ is set, there was an index i where @xs[i] == 0@ and @ys[i] == 1@ and @xs[j] == ys[j]@ for all j < i.
             -- In the same manner, @z1@ and @z2@ won't be updated afterwards.
             update :: (i, i) -> (i, i) -> m (i, i)
             update (z1, z2) (x, y) = do
-                z1' <- newAssigned $ \p -> p z1 + (one - p z1) * (one - p z2) * (one - p y) * p x
-                z2' <- newAssigned $ \p -> p z2 + (one - p z1) * (one - p z2) * (one - p x) * p y
+                -- @f1@ is one if and only if @x > y@ and zero otherwise.
+                -- @(y + 1) `div` (x + 1)@ is zero if and only if @y < x@ regardless of whether @x@ is zero.
+                -- @x@ and @y@ are expected to be of at most @r@ bits where @r << NumberOfBits a@, so @x + 1@ will not be zero either.
+                -- Because of our laws for @finv@, @q // q@ is 1 if @q@ is not zero, and zero otherwise.
+                -- This is exactly the opposite of what @f1@ should be.
+                f1 <- newRanged one $
```

Comment:
`f1` is underconstrained: nothing prevents the adversary from setting `f1` to arbitrary value in range `[0; 1]`

Reply:
(However witness construction is honestly genious)

Reply:
Wait, how do range constraints work, then? `f1` was created with `newRanged`, I thought it will create a record for `f1` in `acRange`, won't it?

Reply:
It will, but this will only constrain `f1` to be in range `[0; 1]`, this does not force it to be a certain value

Reply:
Oh, true, but does it matter? Take a look at `d1` and `d1'`. They have both assigned values based on `f1` and range constraints. If `f1` is set incorrectly, constraints for `d1` and `d1'` won't be satisfied. E.g. if `f1` is zero while `x > y`, `d1` will be set to `y - x` which is outside the range `[0; 2^r]`. Is it still considered unsafe or underconstrained?

Reply:
Oh, wow
Yeah, this works as long $2^{r+1}\le |a|$

Reply:
Also note that this is performed on bits, I think you meant `x` and `y` to be arbitrary field elements so we can avoid `binaryExpansion` in `Ord`

Reply:
We pick `r` in code, so as long as we are careful, it's going to work properly